src/kernel/tracelog.c

```diff
@@ -307,6 +336,78 @@ static void release_tracelog_buffer_locked(tracelog_buffer tb)
     refcount_release(&tb->refcount);
 }
 
+closure_function(5, 1, void, tracelog_file_write_complete,
+                 sg_list, sg, sg_buf, sbg, buffer, b, status_handler, complete, boolean, flushing,
+                 status, s)
+{
+    if (!is_ok(s))
+        msg_err("failed to %s tracelog: %v\n", bound(flushing) ? "flush" : "write to", s);
+    if (bound(flushing)) {
+        async_apply_status_handler(bound(complete), s);
+        closure_finish();
+        return;
+    }
+    deallocate_buffer(bound(b));
+    deallocate_sg_list(bound(sg));
+    if (bound(complete)) {
+        bound(flushing) = true;
+        fsfile_flush(tracelog.logfile, false, (status_handler)closure_self());
+    } else {
+        closure_finish();
+    }
+}
+
+static void tracelog_buffer_fill_locked(buffer b, bytes threshold)
+{
+    list_foreach(&tracelog.collated_entries, l) {
+        tracelog_entry te = struct_from_list(l, tracelog_entry, l);
+        bprintf(b, "[%T, %d, %v", te->t, te->cpu, te->tag);
+        if (te->attrs)
+            bprintf(b, " %v", te->attrs);
+        bprintf(b, "] %s", te->str);
+        list_delete(l);
+        release_tracelog_buffer_locked(te->tb);
+        if (buffer_length(b) > threshold)
+            break;
+    }
+}
+
+/* mutex held */
+static void tracelog_file_write(status_handler complete)
+{
+    tracelog_debug("%s\n", __func__);
+    sg_list sg = allocate_sg_list();
+    if (sg == INVALID_ADDRESS)
+        goto fail_mem;
+    buffer b = allocate_buffer(tracelog.h, TRACELOG_FILE_WRITE_THRESHOLD);
+    tracelog_buffer_fill_locked(b, TRACELOG_FILE_WRITE_THRESHOLD);
+    bytes size = buffer_length(b);
+    sg_buf sgb = sg_list_tail_add(sg, size);
+    if (sgb == INVALID_ADDRESS)
+        goto fail_tail_add;
+    sgb->buf = buffer_ref(b, 0);
+    sgb->size = size;
+    sgb->offset = 0;
+    sgb->refcount = 0;
+    status_handler sh = closure(tracelog.h, tracelog_file_write_complete, sg, sgb, b, complete, false);
+    if (sh == INVALID_ADDRESS)
+        goto fail_mem_sgb;
```

Comment:
Should go to fail_tail_add instead

Reply:
Thanks. I fixed the error handling here and removed the extraneous argument to tracelog_file_write_complete.